src/main/java/org/candlepin/subscriptions/metering/service/prometheus/PrometheusMeteringController.java

```diff
@@ -145,4 +106,159 @@ public void collectOpenshiftMetrics(String account, OffsetDateTime start, Offset
+            );
+        }
+
+        List<Event> events = new LinkedList<>();
+        int totalEvents = 0;
+        for (QueryResultDataResult r : metricData.getData().getResult()) {
+            Map<String, String> metricLabels = r.getMetric();
+            String clusterId = metricLabels.getOrDefault("_id", "");
+            if (!StringUtils.hasText(clusterId)) {
+                // Can't map this metric without a cluster ID. Not likely to happen but we
+                // will log just in case.
+                log.warn("Skipping metric due to missing cluster ID.");
+                continue;
+            }
+
+            for (List<BigDecimal> measurement : r.getValues()) {
+                BigDecimal time = measurement.get(0);
+                BigDecimal value = measurement.get(1);
+
+                if (!isMetricDataValid(labelSets, clusterId, time, value)) {
+                    // Validation already logs the issues. Skip to the next metric value.
+                    continue;
+                }
+
+                // At this point we already know that the cluster labels exist.
+                Map<String, String> clusterLabels = labelSets.get(new LabelSetKey(clusterId, time));
+                Event event = MeteringEventFactory.openShiftClusterCores(
+                    clusterLabels.get("ebs_account"),
+                    clusterLabels.get("_id"),
+                    clusterLabels.get("support"),
+                    clusterLabels.get("usage"),
+                    clock.dateFromUnix(time),
+                    value.doubleValue()
+                );
+                events.add(event);
+                totalEvents++;
+
+                if (events.size() >= openshiftProperties.getEventBatchSize()) {
+                    log.info("Saving {} events", events.size());
+                    eventController.saveAll(events);
+                    events.clear();
+                }
+            }
+        }
+
+        // Flush the remainder
+        if (!events.isEmpty()) {
+            log.info("Saving events: {}", events.size());
+            eventController.saveAll(events);
+        }
+        return totalEvents;
+    }
+
+    private boolean isMetricDataValid(Map<LabelSetKey, Map<String, String>> labelSets,
+        String clusterId, BigDecimal time, BigDecimal value) {
+        if (time == null || value == null) {
+            // Not likely to happen, but will log just in case.
+            log.warn("Skipping metric since time/value pair was invalid: {}/{}", time, value);
+            return false;
+        }
+
+        if (!labelSets.containsKey(new LabelSetKey(clusterId, time))) {
+            // This could potentially happen if the metric existed but the
+            // Labels weren't reported yet.
+            log.warn("Found OpenShift metric but could not associate cluster {} " +
+                "with any labels. Time: {} Value: {}", clusterId, time, value);
+            return false;
+        }
+        return true;
```

Comment:
Curious about everyone's opinion on methods with multiple exit points.

https://hackerchick.com/religious-war-48293-single-vs-multiple/ 

I probably would have wrote this something like this (**DISCLAIMER** completely untested and hastily refactored in notepad*)

```java
    private boolean isMetricDataValid(Map<LabelSetKey, Map<String, String>> labelSets,
        String clusterId, BigDecimal time, BigDecimal value) {

        boolean isTimeValuePairValid = Objects.nonNull(time) && Objects.nonNull(value)

        boolean isMetricExistingAndLabelsNotReported = !labelSets.containsKey(new LabelSetKey(clusterId, time));

        return isTimeValuePairValid && !isMetricExistingAndLabelsNotReported;
     }
```